fem/libceed/operator.hpp

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_LIBCEED_OPERATOR
+#define MFEM_LIBCEED_OPERATOR
+
+#include "../../config/config.hpp"
+
+#ifdef MFEM_USE_CEED
+#include "ceed.hpp"
+#include <ceed.h>
+#include "../../linalg/operator.hpp"
+#include "solvers-utility.h"
+
+namespace mfem
+{
+
+/** A base class to represent a CeedOperator as an MFEM Operator. */
+class MFEMCeedOperator : public Operator
+{
+protected:
+   CeedOperator oper;
+   bool owns_oper;
+   CeedVector u, v;
+
+   MFEMCeedOperator() : oper(nullptr), owns_oper(true), u(nullptr),
+      v(nullptr) { }
+
+public:
+   void Mult(const Vector &x, Vector &y) const;
+   void AddMult(const Vector &x, Vector &y) const;
+   void GetDiagonal(Vector &diag) const;
+   virtual ~MFEMCeedOperator()
+   {
+      if (owns_oper) { CeedOperatorDestroy(&oper); }
+      CeedVectorDestroy(&u);
+      CeedVectorDestroy(&v);
+   }
+   CeedOperator GetCeedOperator() const { return oper; }
+};
+
+class UnconstrainedMFEMCeedOperator : public MFEMCeedOperator
```

Comment:
Should we remove the `MFEM` in `UnconstrainedMFEMCeedOperator` and `MFEMCeedInterpolation`, I added `MFEM` to `MFEMCeedOperator` because it conflicts with the `CeedOperator` defined by libCEED. My favorite option would have been to use a `ceed` namespace below the `mfem` namespace and rename all these objects to `ceed::Operator`, `ceed::UnconstrainedOperator`, `ceed::Interpolator`, and `ceed::VectorContext`...

Reply:
Do you want me to add a `ceed` namespace in this PR or will you do it in #1766 ?

Reply:
I will do it in #1766 , I think it's simpler.